lib/ecs_deploy/instance_fluctuation_manager.rb

```diff
@@ -132,12 +138,51 @@ def extract_target_container_instances(decrease_count, az_to_container_instances
       target_container_instances
     end
 
+    # list tasks whoose desired_status is "STOPPED" but last_status is "RUNNING" in the ECS cluster
+    def list_stopped_running_task_arns
+      stopped_task_arns = ecs_client.list_tasks(cluster: @cluster, desired_status: "STOPPED").flat_map(&:task_arns)
+      stopped_task_arns.each_slice(MAX_DESCRIBABLE_ECS_TASK_COUNT).flat_map do |arns|
+        ecs_client.describe_tasks(cluster: @cluster, tasks: arns).tasks.select do |task|
+          task.desired_status == "STOPPED" && task.last_status == "RUNNING"
+        end
+      end.map(&:task_arn)
+    end
+
+    # list tasks whose desired_status is "RUNNING" or
+    # whoose desired_status is "STOPPED" but last_status is "RUNNING" on the ECS container
+    def list_running_task_arns(container_instance_arn)
+      running_tasks_arn = ecs_client.list_tasks(cluster: @cluster, container_instance: container_instance_arn).flat_map(&:task_arns)
+      stopped_tasks_arn = ecs_client.list_tasks(cluster: @cluster, container_instance: container_instance_arn, desired_status: "STOPPED").flat_map(&:task_arns)
+      stopped_running_task_arns = stopped_tasks_arn.each_slice(MAX_DESCRIBABLE_ECS_TASK_COUNT).flat_map do |arns|
+        ecs_client.describe_tasks(cluster: @cluster, tasks: arns).tasks.select do |task|
+          task.desired_status == "STOPPED" && task.last_status == "RUNNING"
+        end
+      end.map(&:task_arn)
+      running_tasks_arn + stopped_running_task_arns
+    end
+
+    def wait_until_stop_old_tasks(task_arns)
+      @logger.info("All old tasks: #{task_arns.size}")
+      running_tasks = task_arns.each_slice(MAX_DESCRIBABLE_ECS_TASK_COUNT).flat_map do |arns|
+        ecs_client.describe_tasks(cluster: @cluster, tasks: arns).tasks
+      end.select do |task|
+        task.desired_status == "RUNNING" || (task.desired_status == "STOPPED" && task.last_status == "RUNNING")
+      end
+
+      running_tasks.map(&:task_arn).each_slice(MAX_DESCRIBABLE_ECS_TASK_COUNT).each do |arns|
+        ecs_client.wait_until(:tasks_stopped, cluster: @cluster, tasks: arns)
+      end
+      @logger.info("All old tasks are stopped")
+    end
+
     def stop_tasks(arn)
```

Comment:
I think the method name is not proper because this method only stops tasks started by `run_task` or `start_task` and also returns all running tasks. How about calling `list_running_task_arns` outside the method and changing the name?

```ruby
running_task_arns = list_running_task_arns(arn)
stop_onetime_tasks(running_task_arns)
```

Reply:
I agree partially. Why `stop_onetime_tasks`?
What do you mean `onetime`?

Reply:
By `stop_onetime_tasks`, I meant `stop_tasks_not_belonging_service`.

Reply:
Thanks. I got it.